src/symmetric_crypto.cpp

```diff
@@ -0,0 +1,382 @@
+        : _mode{mode}, _secretKey{std::move(secretKey)}
+{
+}
+
+AESCipherBuilder &AESCipherBuilder::setIV(std::vector<uint8_t> const &iv)
+{
+    _iv = iv;
+    return *this;
+}
+
+AESCipherBuilder &AESCipherBuilder::setPadding(SymmetricCipherPadding padding)
+{
+    _padding = padding;
+    return *this;
+}
+
+AESCipherBuilder &AESCipherBuilder::setAuthTag(std::vector<uint8_t> const &authTag)
+{
+    _authTag = authTag;
+    return *this;
+}
+
+bool isAuthenticatedCipherMode(SymmetricCipherMode mode)
+{
+    switch (mode) {
+        case SymmetricCipherMode::GCM:
+            return true;
+        case SymmetricCipherMode::CTR:
+            //[[fallthrough]];
+        case SymmetricCipherMode::CBC:
+            return false;
+    }
+}
+
+std::shared_ptr<AESCipher> AESCipherBuilder::buildEncryptor()
+{
+    if (isAuthenticatedCipherMode(_mode)) {
+        throw MoCOCrWException("Specified cipher supports authenticated encryption."
+                               " buildAuthenticatedEncryptor() should be used instead.");
+    }
+    auto cipher = new AESCipher{_mode, _padding, _secretKey, _iv,
+                                AESCipher::Operation::Encryption};
+    return std::shared_ptr<AESCipher>(cipher);
+}
+
+std::shared_ptr<AESCipher> AESCipherBuilder::buildDecryptor()
+{
+    if (isAuthenticatedCipherMode(_mode)) {
+        throw MoCOCrWException("Specified cipher supports authenticated encryption."
+                               " buildAuthenticatedDecryptor() should be used instead.");
+    }
+    auto cipher = new AESCipher{_mode, _padding, _secretKey, _iv,
+                                AESCipher::Operation::Decryption};
+    return std::shared_ptr<AESCipher>(cipher);
+}
+
+std::shared_ptr<AuthenticatedAESCipher> AESCipherBuilder::buildAuthenticatedEncryptor()
+{
+    if (!isAuthenticatedCipherMode(_mode)) {
+        throw MoCOCrWException("Specified cipher does not support authenticated encryption."
+                               " buildEncryptor() should be used instead.");
+    }
+    auto cipher = new AuthenticatedAESCipher(_mode, _padding, _secretKey, _iv, _authTag,
+                                             AESCipher::Operation::Encryption);
+    return std::shared_ptr<AuthenticatedAESCipher>(cipher);
+}
+
+std::shared_ptr<AuthenticatedAESCipher> AESCipherBuilder::buildAuthenticatedDecryptor()
+{
+    if (!isAuthenticatedCipherMode(_mode)) {
+        throw MoCOCrWException("Specified cipher does not support authenticated encryption."
+                               " buildDecryptor() should be used instead.");
+    }
+    auto cipher = new AuthenticatedAESCipher(_mode, _padding, _secretKey, _iv, _authTag,
+                                             AESCipher::Operation::Decryption);
+    return std::shared_ptr<AuthenticatedAESCipher>(cipher);
+}
+
+
+class QueueOfVectorsMemoryStrategy::Impl {
+public:
+    void write(const std::vector<uint8_t> &block)
+    {
+        _queue.push_back(block);
```

Comment:
I think this should accept a `std::vector<uint8_t> block` and use `_queue.emplace_back(std::move(block));` (or alternatively, accept a `std::vector<uint8_t>&&`).

Reply:
refactored to &&